# Практична: моделювання в озері даних

Конвеєр обробки даних мережі кав’ярень у хмарному сховищі Azure Blob Storage

# Завдання 1. Підключення

## Налаштування та функції

Імпортуємо бібліотеки, задаємо SAS-посилання та функції для роботи з файлами.

In [1]:
# завантаження бібліотек
from azure.storage.blob import ContainerClient
from io import BytesIO
import pandas as pd

In [2]:
# Джерело (тільки читання)
SOURCE_SAS = "https://datalakedatalab.blob.core.windows.net/data-lake?sp=r&st=2026-07-16T08:11:59Z&se=2026-08-31T16:26:59Z&spr=https&sv=2026-02-06&sr=c&sig=JHmTjds1p8Rp%2Bn9nXZUSWSMdkgAkuPdz%2B4KjbAQUlX0%3D"

In [3]:
# Озеро (читання + запис)
LAKE_SAS = "https://datalakedatalab.blob.core.windows.net/students-homework?sp=racwl&st=2026-07-16T08:10:00Z&se=2026-08-31T16:25:00Z&spr=https&sv=2026-02-06&sr=c&sig=wCRv3poEsB8L4%2Bnj3BKZvwcvlX9vGJ5xo93jYaG3no4%3D"

In [4]:
source = ContainerClient.from_container_url(SOURCE_SAS)
lake   = ContainerClient.from_container_url(LAKE_SAS)

STUDENT = "shudria_oleh"
SOURCE_FILES = ["orders_2026_01.json", "orders_2026_02.json", "orders_2026_03.json", "shops.json", "products.json"]

# читання JSON із джерела
def read_json(name):
    raw = source.download_blob(name).readall()
    return pd.read_json(BytesIO(raw))

# читання Parquet зі свого префікса в озері
def read_parquet(path):
    raw = lake.download_blob(f"{STUDENT}/{path}").readall()
    return pd.read_parquet(BytesIO(raw))

# запис DataFrame у Parquet у свій префікс
def write_parquet(df, path):
    buf = BytesIO()
    df.to_parquet(buf, index=False)
    buf.seek(0)
    lake.upload_blob(name=f"{STUDENT}/{path}", data=buf.getvalue(), overwrite=True)

## Перевірка джерела

У SAS-токені джерела є лише дозвіл `sp=r` без права перегляду списку. Тому перевіряємо доступність п’яти відомих файлів за їхніми іменами.

In [5]:
# Перевірка файлів: що лежать у джерелі
for file in SOURCE_FILES:
    props = source.get_blob_client(file).get_blob_properties()
    print(file, props.size, props.last_modified)

orders_2026_01.json 52009 2026-08-11 19:13:07+00:00
orders_2026_02.json 58935 2026-08-11 19:13:07+00:00
orders_2026_03.json 58956 2026-08-11 19:13:07+00:00
shops.json 718 2026-08-11 19:13:07+00:00
products.json 1390 2026-08-11 19:13:07+00:00


## Читання довідників

Читаємо довідники кав’ярень і позицій меню та перевіряємо кількість рядків.

In [6]:
df_source_shops = read_json("shops.json")
df_source_shops

,shop_code,shop_name,city,district,opened_date,seats
0,S01,Kava Hub,Kyiv,Podil,2019-04-15,24
1,S02,Zerno,Kyiv,Pechersk,2021-09-01,16
2,S03,Rynok Coffee,Lviv,Center,2018-06-20,30
3,S04,Bereg,Odesa,Arkadia,2022-05-10,40
4,S05,Steam,Kharkiv,Center,2023-02-01,12


In [7]:
df_source_products = read_json("products.json")
df_source_products

,product_code,product_name,category,size,price
0,P01,Espresso,Coffee,S,45
1,P02,Americano,Coffee,M,55
2,P03,Cappuccino,Coffee,M,70
3,P04,Latte,Coffee,L,80
4,P05,Flat White,Coffee,M,75
5,P06,Green Tea,Tea,M,50
6,P07,Herbal Tea,Tea,M,50
7,P08,Croissant,Bakery,-,60
8,P09,Cinnamon Bun,Bakery,-,65
9,P10,Sandwich,Bakery,-,110


In [8]:
print(f"Довжина shops: {len(df_source_shops)}")
print(f"Довжина products: {len(df_source_products)}")

Довжина shops: 5
Довжина products: 12


# Завдання 2. Bronze

Сира копія джерела без очищення, нормалізації або додавання колонок. Дані зберігаються в тому вигляді, у якому надійшли.

In [9]:
monthly_order_parts = [read_json(f"orders_2026_0{month}.json") for month in (1, 2, 3)]
df_source_orders = pd.concat(monthly_order_parts, ignore_index=True)
print(f"Довжина df_source_orders: {len(df_source_orders)}")

Довжина df_source_orders: 980


## Запис сирих даних

Записуємо об’єднані замовлення та два довідники у форматі Parquet. Завдяки `overwrite=True` повторний запуск перезапише ті самі файли.

In [10]:
write_parquet(df_source_orders, "bronze/orders.parquet")
write_parquet(df_source_shops, "bronze/shops.parquet")
write_parquet(df_source_products, "bronze/products.parquet")

In [11]:
for b in lake.list_blobs(name_starts_with=f"{STUDENT}/bronze"):
    print(b.name, b.size, b.last_modified)

shudria_oleh/bronze/orders.parquet 18580 2026-08-15 16:14:34+00:00
shudria_oleh/bronze/products.parquet 3573 2026-08-15 16:14:34+00:00
shudria_oleh/bronze/readings/readings.parquet 5011 2026-08-12 19:42:46+00:00
shudria_oleh/bronze/sensors/sensors.parquet 4133 2026-08-12 19:42:46+00:00
shudria_oleh/bronze/shops.parquet 3974 2026-08-15 16:14:34+00:00
shudria_oleh/bronze/stations/stations.parquet 4814 2026-08-12 19:42:46+00:00


Вкладені таблиці `stations`, `sensors` і `readings` залишилися від попередньої практичної. У цій роботі використовуємо три файли, що лежать безпосередньо в `bronze/`.

# Завдання 3. Silver

Читаємо замовлення тільки з Bronze. Спочатку перевіряємо якість даних, а потім очищаємо їх.

In [12]:
df_bronze_orders = read_parquet("bronze/orders.parquet")
df_bronze_orders

,order_id,order_ts,shop_code,product_code,qty,unit_price,payment_type
0,O500259,2026-01-27 12:05:00,S04,P04,2.0,80,Card
1,O500252,2026-01-03 09:30:00,S05,P04,2.0,80,Card
2,O500257,2026-01-30 11:35:00,S04,P11,1.0,95,Cash
3,O500127,2026-01-03 09:40:00,S01,P10,3.0,110,Card
4,O500225,2026-01-28 17:35:00,S03,P04,3.0,80,Cash
...,...,...,...,...,...,...,...
975,O500741,2026-03-19 07:30:00,S04,P01,1.0,45,Card
976,O500781,2026-03-04 09:15:00,S02,P08,1.0,60,Card
977,O500765,2026-03-15 09:30:00,S02,P04,1.0,80,Card
978,O500872,2026-03-05 10:20:00,S03,P09,1.0,65,Cash


## Профілювання якості замовлень

Перевіряємо дублікати, різні варіанти написання значень і пропуски в `qty`. Поки що дані не змінюємо.

In [13]:
df_bronze_orders.describe(include="all")

,order_id,order_ts,shop_code,product_code,qty,unit_price,payment_type
count,980,980,980,980,944.000000,980.000000,980
unique,960,928,10,12,NaN,NaN,4
top,O500240,2026-01-04 10:30:00,S03,P03,NaN,NaN,Cash
freq,2,4,204,148,NaN,NaN,461
mean,NaN,NaN,NaN,NaN,1.355932,69.979592,NaN
std,NaN,NaN,NaN,NaN,0.584718,17.557714,NaN
min,NaN,NaN,NaN,NaN,1.000000,45.000000,NaN
25%,NaN,NaN,NaN,NaN,1.000000,55.000000,NaN
50%,NaN,NaN,NaN,NaN,1.000000,70.000000,NaN
75%,NaN,NaN,NaN,NaN,2.000000,80.000000,NaN


In [14]:
df_bronze_orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 980 entries, 0 to 979
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   order_id      980 non-null    str    
 1   order_ts      980 non-null    str    
 2   shop_code     980 non-null    str    
 3   product_code  980 non-null    str    
 4   qty           944 non-null    float64
 5   unit_price    980 non-null    int64  
 6   payment_type  980 non-null    str    
dtypes: float64(1), int64(1), str(5)
memory usage: 88.4 KB


In [15]:
print("Повні дублікати:", df_bronze_orders.duplicated().sum())
print("Унікальні shop_code:", df_bronze_orders['shop_code'].nunique())
print("Унікальні payment_type:", df_bronze_orders['payment_type'].nunique())
print("Порожні qty:", df_bronze_orders['qty'].isnull().sum())

Повні дублікати: 20
Унікальні shop_code: 10
Унікальні payment_type: 4
Порожні qty: 36


## Видалення повних дублікатів

Створюємо окремий DataFrame для Silver і видаляємо повністю однакові рядки. Дані в Bronze не змінюємо.

In [16]:
df_silver_orders = df_bronze_orders.drop_duplicates().reset_index(drop=True)
print(f"до видалення дублікатів: {len(df_bronze_orders)}")
print(f"після видалення дублікатів: {len(df_silver_orders)}")
print(f"перевірка дублікатів після видалення: {df_silver_orders.duplicated().sum()}")

до видалення дублікатів: 980
після видалення дублікатів: 960
перевірка дублікатів після видалення: 0


## Нормалізація коду кав’ярні

Прибираємо пробіли на початку та в кінці `shop_code` і переводимо значення у верхній регістр.

Після нормалізації мають залишитися п’ять унікальних кодів: `S01`, `S02`, `S03`, `S04`, `S05`.

In [17]:
df_silver_orders["shop_code"] = df_silver_orders["shop_code"].str.strip().str.upper()

df_silver_orders["shop_code"].unique()

<ArrowStringArray>
['S04', 'S05', 'S01', 'S03', 'S02']
Length: 5, dtype: str

## Нормалізація способу оплати

Спочатку переглядаємо унікальні значення `payment_type`.

Після нормалізації мають залишитися два значення: `Card` і `Cash`.

In [18]:
df_silver_orders["payment_type"].unique()

<ArrowStringArray>
['Card', 'Cash', 'CASH', 'CARD']
Length: 4, dtype: str

In [19]:
df_silver_orders["payment_type"] = df_silver_orders["payment_type"].str.capitalize()

df_silver_orders["payment_type"].unique()

<ArrowStringArray>
['Card', 'Cash']
Length: 2, dtype: str

## Заповнення пропусків у кількості товарів

У колонці `qty` є пропуски. За умовою замінюємо їх на `1`.

Після заповнення переводимо колонку в цілий тип.

In [20]:
df_silver_orders["qty"] = df_silver_orders["qty"].fillna(1).astype(int)

## Перетворення дати й часу замовлення

Перетворюємо `order_ts` із тексту на тип `datetime`.

З `order_ts` створюємо `order_date` без часу. Ця колонка знадобиться для групування продажів за днями.

In [21]:
df_silver_orders["order_ts"] = pd.to_datetime(df_silver_orders["order_ts"], format="%Y-%m-%d %H:%M:%S")

df_silver_orders["order_date"] = df_silver_orders["order_ts"].dt.date

## Розрахунок суми продажу

Обчислюємо суму кожної позиції замовлення:

`amount = qty × unit_price`

Беремо `unit_price` із замовлення, тому що це ціна товару на момент продажу.

In [22]:
df_silver_orders["amount"] = df_silver_orders["qty"] * df_silver_orders["unit_price"]

## Перевірка очищених замовлень

Переглядаємо очищені дані, кількість рядків і типи колонок.

У Silver має залишитися 960 рядків. `qty` повинен мати цілий тип, `order_ts` тип дати й часу, а в `order_date` та `amount` не повинно бути пропусків.

In [23]:
df_silver_orders

,order_id,order_ts,shop_code,product_code,qty,unit_price,payment_type,order_date,amount
0,O500259,2026-01-27 12:05:00,S04,P04,2,80,Card,2026-01-27,160
1,O500252,2026-01-03 09:30:00,S05,P04,2,80,Card,2026-01-03,160
2,O500257,2026-01-30 11:35:00,S04,P11,1,95,Cash,2026-01-30,95
3,O500127,2026-01-03 09:40:00,S01,P10,3,110,Card,2026-01-03,330
4,O500225,2026-01-28 17:35:00,S03,P04,3,80,Cash,2026-01-28,240
...,...,...,...,...,...,...,...,...,...
955,O500741,2026-03-19 07:30:00,S04,P01,1,45,Card,2026-03-19,45
956,O500781,2026-03-04 09:15:00,S02,P08,1,60,Card,2026-03-04,60
957,O500765,2026-03-15 09:30:00,S02,P04,1,80,Card,2026-03-15,80
958,O500872,2026-03-05 10:20:00,S03,P09,1,65,Cash,2026-03-05,65


In [24]:
df_silver_orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 960 entries, 0 to 959
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   order_id      960 non-null    str           
 1   order_ts      960 non-null    datetime64[us]
 2   shop_code     960 non-null    str           
 3   product_code  960 non-null    str           
 4   qty           960 non-null    int64         
 5   unit_price    960 non-null    int64         
 6   payment_type  960 non-null    str           
 7   order_date    960 non-null    object        
 8   amount        960 non-null    int64         
dtypes: datetime64[us](1), int64(3), object(1), str(4)
memory usage: 84.0+ KB


## Запис очищених даних у Silver

Записуємо очищені замовлення в Silver. Довідники кав’ярень і позицій меню читаємо з Bronze та переносимо без змін.

In [25]:
print(f"Довжина df_silver_orders: {len(df_silver_orders)}")

df_silver_shops = read_parquet("bronze/shops.parquet")
print(f"Довжина df_silver_shops: {len(df_silver_shops)}")

df_silver_products = read_parquet("bronze/products.parquet")
print(f"Довжина df_silver_products: {len(df_silver_products)}")

Довжина df_silver_orders: 960
Довжина df_silver_shops: 5
Довжина df_silver_products: 12


In [26]:
write_parquet(df_silver_orders, "silver/orders.parquet")

write_parquet(df_silver_shops, "silver/shops.parquet")

write_parquet(df_silver_products, "silver/products.parquet")

# Завдання 4. Gold

Будуємо зоряну схему з двох вимірів і двох таблиць фактів. Gold читає дані виключно із Silver.

## Вимір кав’ярень `dim_shop`

Один рядок виміру відповідає одній кав’ярні. Сортуємо записи за бізнес-ключем `shop_code` і додаємо сурогатний ключ `shop_key`.

In [27]:
dim_shop = read_parquet("silver/shops.parquet")

dim_shop = dim_shop.sort_values("shop_code").reset_index(drop=True)

dim_shop.insert(0, "shop_key", range(1, len(dim_shop) + 1))

dim_shop

,shop_key,shop_code,shop_name,city,district,opened_date,seats
0,1,S01,Kava Hub,Kyiv,Podil,2019-04-15,24
1,2,S02,Zerno,Kyiv,Pechersk,2021-09-01,16
2,3,S03,Rynok Coffee,Lviv,Center,2018-06-20,30
3,4,S04,Bereg,Odesa,Arkadia,2022-05-10,40
4,5,S05,Steam,Kharkiv,Center,2023-02-01,12


## Вимір позицій з меню `dim_product`

Один рядок виміру відповідає одній позиції з меню. Сортуємо записи за бізнес-ключем `product_code` і додаємо сурогатний ключ `product_key`.

In [28]:
dim_product = read_parquet("silver/products.parquet")

dim_product = dim_product.sort_values("product_code").reset_index(drop=True)

dim_product.insert(0, "product_key", range(1, len(dim_product) + 1))

dim_product

,product_key,product_code,product_name,category,size,price
0,1,P01,Espresso,Coffee,S,45
1,2,P02,Americano,Coffee,M,55
2,3,P03,Cappuccino,Coffee,M,70
3,4,P04,Latte,Coffee,L,80
4,5,P05,Flat White,Coffee,M,75
5,6,P06,Green Tea,Tea,M,50
6,7,P07,Herbal Tea,Tea,M,50
7,8,P08,Croissant,Bakery,-,60
8,9,P09,Cinnamon Bun,Bakery,-,65
9,10,P10,Sandwich,Bakery,-,110


## Таблиця фактів `fact_sales`

Читаємо очищені замовлення із Silver та приєднуємо до них сурогатні ключі вимірів.

Коди `shop_code` і `product_code` використовуємо для пошуку відповідних ключів, але у фінальній таблиці фактів залишаємо тільки `shop_key` і `product_key`.

Один рядок `fact_sales` відповідає одній позиції замовлення.

In [29]:
fact_sales = read_parquet("silver/orders.parquet")

print(f"Довжина fact_sales до об'єднання: {len(fact_sales)}")

fact_sales = fact_sales.merge(dim_shop[["shop_key", "shop_code"]], on="shop_code", how="left")

fact_sales = fact_sales.merge(dim_product[["product_key", "product_code"]], on="product_code", how="left")

print(f"Довжина fact_sales після об'єднання: {len(fact_sales)}")
print(f"Кількість пропущених значень у fact_sales:\n{fact_sales.isnull().sum()}")

fact_sales = fact_sales[["order_id", "shop_key", "product_key", "order_date", "qty", "unit_price", "amount", "payment_type"]]

fact_sales

Довжина fact_sales до об'єднання: 960
Довжина fact_sales після об'єднання: 960
Кількість пропущених значень у fact_sales:
order_id        0
order_ts        0
shop_code       0
product_code    0
qty             0
unit_price      0
payment_type    0
order_date      0
amount          0
shop_key        0
product_key     0
dtype: int64


,order_id,shop_key,product_key,order_date,qty,unit_price,amount,payment_type
0,O500259,4,4,2026-01-27,2,80,160,Card
1,O500252,5,4,2026-01-03,2,80,160,Card
2,O500257,4,11,2026-01-30,1,95,95,Cash
3,O500127,1,10,2026-01-03,3,110,330,Card
4,O500225,3,4,2026-01-28,3,80,240,Cash
...,...,...,...,...,...,...,...,...
955,O500741,4,1,2026-03-19,1,45,45,Card
956,O500781,2,8,2026-03-04,1,60,60,Card
957,O500765,2,4,2026-03-15,1,80,80,Card
958,O500872,3,9,2026-03-05,1,65,65,Cash


## Денна таблиця фактів `fact_daily_sales`

Групуємо продажі за кав’ярнею та датою.

Групуємо `fact_sales` за `shop_key` і `order_date` та розраховуємо:

- `orders_count`: кількість унікальних замовлень
- `items_sold`: кількість проданих одиниць
- `revenue`: загальний виторг

Один рядок `fact_daily_sales` відповідає результатам однієї кав’ярні за один день.

In [30]:
fact_daily_sales = fact_sales.groupby(["shop_key", "order_date"]).agg(
    orders_count=pd.NamedAgg(column="order_id", aggfunc="nunique"),
    items_sold=pd.NamedAgg(column="qty", aggfunc="sum"),
    revenue=pd.NamedAgg(column="amount", aggfunc="sum")
).reset_index()

fact_daily_sales

,shop_key,order_date,orders_count,items_sold,revenue
0,1,2026-01-01,2,3,235
1,1,2026-01-02,1,1,95
2,1,2026-01-03,6,9,785
3,1,2026-01-04,2,3,225
4,1,2026-01-05,2,3,170
...,...,...,...,...,...
391,5,2026-03-27,3,4,280
392,5,2026-03-28,1,1,85
393,5,2026-03-29,3,4,225
394,5,2026-03-30,1,2,100


## Запис таблиць Gold

Зберігаємо два виміри та дві таблиці фактів у зоні Gold.

In [31]:
write_parquet(dim_shop, "gold/dim_shop.parquet")

write_parquet(dim_product, "gold/dim_product.parquet")

write_parquet(fact_sales, "gold/fact_sales.parquet")

write_parquet(fact_daily_sales, "gold/fact_daily_sales.parquet")

# Завдання 5. Перевірка та аналітика

## 5.1. Перевірка кількості рядків

Повторно читаємо чотири таблиці з Gold і перевіряємо кількість рядків.

Очікувані результати:

- `dim_shop`: 5 кав’ярень
- `dim_product`: 12 позицій меню
- `fact_sales`: 960 позицій продажу
- `fact_daily_sales`: 396 комбінацій кав’ярні та дати

In [32]:
dim_shop_gold = read_parquet("gold/dim_shop.parquet")
dim_product_gold = read_parquet("gold/dim_product.parquet")
fact_sales_gold = read_parquet("gold/fact_sales.parquet")
fact_daily_sales_gold = read_parquet("gold/fact_daily_sales.parquet")

print(f"Довжина dim_shop_gold: {len(dim_shop_gold)}")
print(f"Довжина dim_product_gold: {len(dim_product_gold)}")
print(f"Довжина fact_sales_gold: {len(fact_sales_gold)}")
print(f"Довжина fact_daily_sales_gold: {len(fact_daily_sales_gold)}")

Довжина dim_shop_gold: 5
Довжина dim_product_gold: 12
Довжина fact_sales_gold: 960
Довжина fact_daily_sales_gold: 396


## 5.2. Перевірка посилальної цілісності

Перевіряємо, що кожен `shop_key` із таблиці фактів є в `dim_shop`, а кожен `product_key` є в `dim_product`.

Обидві перевірки мають повернути `True`.

In [33]:
print(f"Перевірка shop_key: {fact_sales_gold['shop_key'].isin(dim_shop_gold['shop_key']).all()}")
print(f"Перевірка product_key: {fact_sales_gold['product_key'].isin(dim_product_gold['product_key']).all()}")

Перевірка shop_key: True
Перевірка product_key: True


## 5.3. Звірка загального виторгу

Порівнюємо суму `amount` у детальній таблиці `fact_sales` із сумою `revenue` у денній таблиці `fact_daily_sales`.

Суми мають бути однаковими.

In [34]:
revenue_before = fact_sales_gold["amount"].sum()
revenue_after = fact_daily_sales_gold["revenue"].sum()

print(f"Перевірка amount = revenue: {revenue_before == revenue_after}")

Перевірка amount = revenue: True


## 5.4. Виторг за категоріями меню

Приєднуємо до продажів категорію товару з `dim_product`, групуємо дані за `category` та обчислюємо сумарний виторг.

Сортуємо результат за спаданням виторгу.

In [35]:
fact_sales_product = fact_sales_gold.merge(dim_product_gold[['product_key', 'category']], on="product_key", how="left")

fact_sales_product.groupby("category").agg(
    виторг = pd.NamedAgg(column="amount", aggfunc="sum")).reset_index().sort_values("виторг", ascending=False)

,category,виторг
1,Coffee,49915
0,Bakery,19955
2,Dessert,14850
3,Tea,5600


## 5.5. Виторг за кав’ярнями

Приєднуємо назви кав’ярень із `dim_shop` до таблиці продажів, групуємо за `shop_name` та обчислюємо сумарний виторг кожної кав’ярні.

Сортуємо результат за спаданням виторгу.

In [36]:
fact_sales_shop = fact_sales_gold.merge(dim_shop_gold[['shop_key', 'shop_name']], on="shop_key", how="left")

fact_sales_shop.groupby("shop_name").agg(
    виторг = pd.NamedAgg(column="amount", aggfunc="sum")).reset_index().sort_values("виторг", ascending=False)

,shop_name,виторг
2,Rynok Coffee,20075
1,Kava Hub,19290
3,Steam,18275
0,Bereg,16750
4,Zerno,15930


# Висновок

У цій роботі я побудував конвеєр обробки даних із шарами Bronze, Silver і Gold.

У Bronze записав 980 сирих позицій замовлень без змін. У Silver видалив 20 повних дублікатів, привів до одного вигляду коди кав’ярень і способи оплати, заповнив пропуски в `qty`, додав дату та суму продажу. Після очищення залишилося 960 рядків.

У Gold створив виміри `dim_shop` і `dim_product`, а також таблиці `fact_sales` і `fact_daily_sales`. Перевірка ключів повернула `True`, а сума виторгу в обох таблицях фактів дорівнює 90320.

Найбільший виторг серед категорій має `Coffee` із сумою 49915. Серед кав’ярень перше місце посідає `Rynok Coffee` із виторгом 20075.